### ISSUE TITLE
"Required input 'Signup' not provided" on setAccountPassword

### ISSUE THREAD

**Issue by gmonteeeiro** [gmonteeeiro/twitterbot#2]
After send the Twitter verification code the method setAccountPassword is receiving this response
```
{
    "errors": [
        {
            "code": 366,
            "message": "Required input 'Signup' not provided."
        }
    ]
}
```

But using the creation Twitter account web page, the request don't have this 'Signup' input:

![image](https://user-images.githubusercontent.com/26847545/79507749-a129f380-800e-11ea-9713-901081ffcfbc.png)

**Comment by ceres-c** [comment]
I've stumbled upon this issue while I was developing my own _hack it together in python in as little time as possible_™ version of this project. This is my current code and I'm sending device events as well to their logging api
https://gist.github.com/ceres-c/7c16a40c10cb476cce2c4b9023341377

I've checked the request flow with Burp and it seems to be pretty consistent with a legit registration, but since I face the same error as you do, I guess something's still off. I'm posting this here because it might be a starting point for someone else to pick this up, I guess

**Comment by Aze1337** [comment]
Did you guys figure it out ? I just got it, I'll investigate and update here if any of you is interested.

**Comment by ceres-c** [comment]
I haven't been looking into this since the day I posted the above message. Good luck though, I'd be curious to find out a proper fix

**Comment by MrFatoni** [comment]
change last number at flow_token, every step added by 1, that's it

**Comment by ceres-c** [comment]
LOL
I haven't tried it, but thanks

**Comment by Aze1337** [comment]
> change last number at flow_token, every step added by 1, that's it

rofl would of been funny if that was it but I don't see how it makes sense ?

look
![image](https://user-images.githubusercontent.com/76887650/112738240-7a22b400-8f37-11eb-8ae4-33c3872236df.png)
the last request below uses the flow_token that was scraped in the previous request (incremented 2) , never went to :1
![image](https://user-images.githubusercontent.com/76887650/112738254-90307480-8f37-11eb-9d2c-a46c01c92f66.png)
they match..

**Comment by Aze1337** [comment]
the only think I'm missing in my ChoosePassword last request would be the x-csrf-token. anything else is there as I found where they come from but I still get the same error as OP. I know that if you run it through a browser, it would be sent as cookie named "ct0" but I haven't found yet what javascript code does it uses to generate it... do you guys have figured it out? (obviously without a browser, I want it all http reqs) .. let me know

**Comment by ceres-c** [comment]
Python requests sessions automatically save cookies in a cookies dict
```client_event_headers_POST["x-csrf-token"]	= session.cookies['ct0']```

**Comment by Aze1337** [comment]
> Python requests sessions automatically save cookies in a cookies dict when it's pushed down by the server
> `client_event_headers_POST["x-csrf-token"] = session.cookies['ct0']`

thanks for answer. can you clarify which endpoint would I get the response as Cookie ? I can't seems to find a response with it, it just randomly comes as header & cookie when using browser but no request seems to send it on my end.

**Comment by ceres-c** [comment]
I don't remember, sorry, I haven't been touching this code for months. Take a look at my code linked in the above message, it should be commented

**Comment by Aze1337** [comment]
sadly I think your solution uses web browser ? (driver.get) to load google and then go on twitter so it uses the browser first then continue with http browser ? I'm looking to figure out with http reqs only how does the browser generate it ;/
i might read your code wrong if thats not the case
thanks tho

**Comment by ceres-c** [comment]
Nope, my code uses selenium just to solve the javascript challenge which is pushed down by twitter and requires DOM objects to be solved, so you can't use libs like js2py and alike as they do not provide a DOM environment

**Comment by Aze1337** [comment]
what was the javascript challenge ? I could go past all the http requests without it until the last one . I cant give up till I find a way haha

its fine if you dont want to spend time on it anymore! i checked all endpoints at the top of the code and none seems to send me back a cookie/header with the csrf/ct0 as response

i checked twitter requests a bunch of times and it's never sent so I'm almost sure as my first idea is - some javascript code is executed on web browser and then it's set as cookie/header. but I'm unsure which code generates it from the .js files twitter have.

i'll try to figure it out and update here if i can help someone in the future who needs it.
have a good day/night

**Comment by ceres-c** [comment]
On password update you are required to send the response to a challenge which is calculated in javascript. That js requires DOM objects and changes every time. If you don't send it on password request, the action is denied

**Comment by Aze1337** [comment]
Oh I see ! - https://twitter.com/i/js_inst?c_name=ui_metrics - those changes all the time, but you can still scrape them with http req and submit them until the last request which is like the original poster - refused for some reasons- I believe I'm missing the x-csrf-token way to generate but I'll know tomorrow most likely if that was the issue as I'm going to scrape everything with web browser and then run through http reqs- unless I figure out how it works without browser. hopefully will soon. :(

**Comment by Aze1337** [comment]
people who wants to know how I scraped them, here my way with c#. was pretty annoying not gonna lie
![image](https://user-images.githubusercontent.com/76887650/112740169-b4488180-8f48-11eb-90e9-943663588820.png)

**Comment by MrFatoni** [comment]
[https://github.com/MrFatoni/test-twitter-creator/blob/main/twt.php](https://github.com/MrFatoni/test-twitter-creator/blob/main/twt.php)

try this, working fine

**Comment by gmonteeeiro** [comment]
thanks @MrFatoni, i'll see your code

**Comment by koala9712** [comment]
Slightly old issue but might be worth updating.

I was getting this error upon submitting the verification code (phone/email) and I solved it by passing in the exactly same payload as through the browser- what I was doing wrong was, it's differnet for phone and email 💀

Anyways, I also saw that it doesn't give a damn if the js challenge was solved or not, heck, I completely removed js_instrumentations from the payload and it still gave me the same error!

Second anyways, that solved, I am now getting a new error saying `{"errors":[{"code":366,"message":"Missing data."}]}`- any thoughts on solving this? Much appreciated!

**Comment by koala9712** [comment]
I think it's worth noting I'm not doing any of the client_event requests.

**Comment by cheffym8** [comment]
> 

can u show ur payload

**Comment by cheffym8** [comment]
> Slightly old issue but might be worth updating.
> 
> I was getting this error upon submitting the verification code (phone/email) and I solved it by passing in the exactly same payload as through the browser- what I was doing wrong was, it's differnet for phone and email 💀
> 
> Anyways, I also saw that it doesn't give a damn if the js challenge was solved or not, heck, I completely removed js_instrumentations from the payload and it still gave me the same error!
> 
> Second anyways, that solved, I am now getting a new error saying `{"errors":[{"code":366,"message":"Missing data."}]}`- any thoughts on solving this? Much appreciated!

can u show ur payload

**Comment by koala9712** [comment]
https://mystb.in/ElectoralLipsFound.perl
Note:
- This is for when you enter the otp
- The request to ask for verification code, it's payload also changes for phone/email

**Comment by cheffym8** [comment]
> https://mystb.in/ElectoralLipsFound.perl Note:
> 
> * This is for when you enter the otp
> * The request to ask for verification code, it's payload also changes for phone/email

can u also show ur payload for the password request please, thats where im getting the error and looks 1:1 of browser

**Comment by koala9712** [comment]
Check out the [repo](https://github.com/koala9712/Twitter-Account-Generator) :)

**Comment by cheffym8** [comment]
> Check out the [repo](https://github.com/koala9712/Twitter-Account-Generator) :)

its private :(
